### ISSUE TITLE
@azure/msal-node-extensions doesn't have pre-compiled binaries

### ISSUE THREAD

**Issue by sadasant** [AzureAD/microsoft-authentication-library-for-js#3332]
## Library
- [x] `@azure/msal-node-extensions@1.0.0-alpha.6`

## Description

`@azure/msal-node-extensions@1.0.0-alpha.6` builds its binaries when the package gets installed by a developer. As the readme says, this means that developers will need to have `node-gyp`, which in turn has considerable requirements on each platform. For example, on Windows, `node-gyp` requires Visual Studio with C++, which means `@azure/msal-node-extensions@1.0.0-alpha.6` is requiring Windows users to install a separate dependency that weighs several gigabytes on its own.

Another issue of not having pre-compiled binaries is that if a developer installs `@azure/msal-node-extensions@1.0.0-alpha.6` on Node 14 and then they switch to a different version of Node (let's say, Node 15), once they try to use persistence they get an error similar to the following:

```
Error: The module '/home/user/workspace/azure-sdk-for-js/common/temp/node_modules/.pnpm/@azure/msal-node-extensions@1.0.0-alpha.6/node_modules/@azure/msal-node-extensions/build/Release/dpapi.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 88. This version of Node.js requires
NODE_MODULE_VERSION 83. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
```

Many users won't be able to consider this dependency while it requires this setup. Ideally the binaries would come pre-compiled, specially for developers using this library on production environments.

## Source

- [x] Internal (Microsoft)
- [ ] Customer request

**Comment by sameerag** [comment]
Thanks @sadasant for raising this.

We evaluated this request with the team and also checked [keytar](https://www.npmjs.com/package/keytar)’s model of shipping the binaries. We decided we do not want to go that route – let’s take keytar as an example: they ship [~90 binaries](https://github.com/atom/node-keytar/releases) per release today which is something we will not be able to sustain considering the OS and node version combos we may have to do this for.
 
We support node versions 10|12|14 (and any future stable [Node.js LTE versions](https://nodejs.org/en/about/releases/)), and provide a work around if any app developer would want to bypass the engines check we placed in our node package. A PR documenting this behavior is in progress [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/3290). Similar check will be soon added to msal-node-extensions.

I am still keeping this open to evaluate any other alternatives here. Can you also please raise an ADO ticket as discussed offline so we can track it internally?

**Comment by sameerag** [comment]
@sadasant  We evaluated this request with the team and also checked keytar’s model of shipping the binaries. We decided we do not want to go that route – let’s take [keytar](https://www.npmjs.com/package/keytar) as an example: they ship [~90 binaries](https://github.com/atom/node-keytar/releases) per release today which is something we will not be able to sustain considering the OS and node version combos we may have to do this for.
 
We support node versions 10|12|14 (and any future stable [Node.js LTE versions](https://nodejs.org/en/about/releases/)), and provide a work around if any app developer would want to bypass the engines check we placed in our node [package](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/package.json). A PR documenting this behavior can be referenced [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/3290). Similar check will be soon added to msal-node-extensions.
 
Note: Can you please help us by creating an ADO item to track this ask? We are moving all our feature requests to ADO and it will help the team to route the requests there.

One consideration I haven't yet thought through is to provide a utility for the customers to generate their own binaries as an optional command that can be run on an installed package - I haven't yet set aside time for this and but will keep you posted as we have any update. 

cc @jasonnutter

**Comment by sameerag** [comment]
@sadasant Since you folks decided to solve this issue differently, is it alright to close this issue? Please let me know.

**Comment by sameerag** [comment]
Closing it. Please open a new issue if need be.

**Comment by sinedied** [comment]
@sameerag Could this issue be re-opened? The original problem you're citing above about having to build ~90 binaries is no more relevant when using Node N-API, as you can see that Keytar now supports all versions and platforms with only 11 prebuilt binaries: https://github.com/atom/node-keytar/releases/tag/v7.9.0

This is also linked to a blocking issue for us, mentioned here: https://github.com/Azure/azure-sdk-for-js/issues/21195

**Comment by jasonnutter** [comment]
Reopening, as this is being requested again. As a workaround, I would suggest consuming applications/libraries to mark MSAL Node Extensions (or its dependent) as an optional dependency (and handle its potential absence), and/or to precompile your application (e.g. using pkg).

**Comment by scottdurow** [comment]
Hi @jasonnutter - I have a package that is run using npx/npm exec - so I don't think pkg is an option here?

Are there any options to distribute the binaries so a package command can be run using npx without the need for the compilation step?

I really want to use the node-extensions since they are perfect for authentication across node npx commands - but not if you have to have all the compilations pre-reqs installed.

**Comment by sameerag** [comment]
cc @tnorling

**Comment by tnorling** [comment]
Addressed in #6084 and will be included in the next release of msal-node-extensions under the beta tag.
